File: 17. Ch-4-3.java
import java.util.*;
class Node {
    int data;
    Node left, right;

    Node(int data) {
        this.data = data;
        this.left = null;
        this.right = null;
    }
}

class BuildTree {
    int idx = 0;
    Node buildTree(int[] arr) {
        if (idx >= arr.length || arr[idx] == -1) {
            idx++;
            return null;
        }
        Node root = new Node(arr[idx++]);
        root.left = buildTree(arr);
        root.right = buildTree(arr);
        return root;
    }
}

public class Main {
    static boolean isMirror(Node root1, Node root2) {
        if (root1 == null && root2 == null)
            return true;
        if (root1 == null || root2 == null)
            return false;
        return (root1.data == root2.data)
                && isMirror(root1.left, root2.right)
                && isMirror(root1.right, root2.left);
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        String[] input1 = sc.nextLine().split(" ");
        int[] arr1 = new int[input1.length];
        for (int i = 0; i < input1.length; i++)
            arr1[i] = Integer.parseInt(input1[i]);

        String[] input2 = sc.nextLine().split(" ");
        int[] arr2 = new int[input2.length];
        for (int i = 0; i < input2.length; i++)
            arr2[i] = Integer.parseInt(input2[i]);

        BuildTree bt1 = new BuildTree();
        BuildTree bt2 = new BuildTree();

        Node root1 = bt1.buildTree(arr1);
        Node root2 = bt2.buildTree(arr2);
        if (isMirror(root1, root2)) {
            System.out.println("The two trees are mirrors of each other.");
        } else {
            System.out.println("The two trees are not mirrors of each other.");
        }
    }
}

File: 17.Ch-4-3.py
class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

class BuildTree:
    def __init__(self):
        self.idx = 0
    def build_tree(self, arr):
        if self.idx >= len(arr):
            return None
        if arr[self.idx] == -1:
            self.idx += 1
            return None
        root = Node(arr[self.idx])
        self.idx += 1
        root.left = self.build_tree(arr)
        root.right = self.build_tree(arr)
        return root
#Main__________________
arr1 = list(map(int, input().split()))
arr2 = list(map(int, input().split()))
bt1 = BuildTree()
bt2 = BuildTree()
root1 = bt1.build_tree(arr1)
root2 = bt2.build_tree(arr2)
def is_mirror(root1, root2):
    if root1 is None and root2 is None:
        return True
    if root1 is None or root2 is None:
        return False
    if root1.data != root2.data:
        return False
    return (is_mirror(root1.left, root2.right) and
            is_mirror(root1.right, root2.left))

if is_mirror(root1, root2):
    print("The two trees are mirrors of each other.")
else:
    print("The two trees are not mirrors of each other.")
